projs/shadow/shadow-engine/core/inc/shadow/core/jobs/Job.h

```diff
@@ -0,0 +1,213 @@
+         */
+        inline void Execute(size_t firstQueue) {
+            TaskGroup task;
+            for (size_t i = 0; i < nThreads; i++) {
+                TaskQueue& tasks = queues[firstQueue & nThreads];
+                while (tasks.Pop(task))
+                    task.Execute();
+                firstQueue++;
+            }
+        }
+    };
+
+    /**
+     * Initialize the Task management system.
+     * @param maxThreads the maximum number of threads that can be dedicated entirely to running Tasks.
+     */
+    void Init(uint32_t maxThreads = 0xFFFFFFFF);
+
+    /**
+     * Prepare the Task management system for engine shutdown.
+     */
+    void Destroy();
+
+    /**
+     * Fetch the number of Threads currently running, for the given priority.
+     * @param p the priority of threads to check
+     * @return The number of threads currently running a task with the given priority.
+     */
+    size_t GetThreadsOfPriority(Priority p = Priority::HIGHEST);
+
+    /**
+     * Run the given task off-thread, with the given priority.
+     * The function will run, and then return. The task cannot be restarted, unless you call Run again.
+     * @param context a task context, with priority
+     * @param task the function to run off-thread
+     */
+    void Run(ExecutionContext& context, const std::function<void(TaskArguments)>& task);
+
+    /**
+     * Dispatch a set of jobs, running the same code with different arguments.
+     * The arguments that change are the task index and group index.
+     * Usually use this where iterating over a static list and performing an expensive task on every object within.
+     * @param context a task context, with priority
+     * @param jobs the number of jobs to create, total. Usually, max index of the list being processed
+     * @param groups the number of jobs to run per thread. Usually, jobs / max_threads if jobs > max_threads, or 5 otherwise.
+     * @param task the function to run per index
+     * @param sharedMem the size of a shared memory buffer to allocate for every group.
+     */
+    void Dispatch(ExecutionContext& context, size_t jobs, size_t groups, const std::function<void(TaskArguments)>& task, size_t sharedMem = 0);
+
+    /**
+     * Fetch the number of groups to be created for a given number of jobs, for groups of a given size.
+     * @param jobs the total number of jobs to create
+     * @param groups the number of jobs to put in each group
+     * @return The number of groups to create to fit all jobs
+     */
+    size_t DispatchGroups(size_t jobs, size_t groups);
+
+    /**
+     * Check whether any threads are working for the given context.
+     * @param context a task context, with priority
+     * @return Whether any threads are working for this context
+     */
+    bool IsWorking(const ExecutionContext& context);
+
+    /**
+     * Wait for the given context to stop running.
+     * The current thread will be added to the thread pool for jobs.
```

Comment:
It means that if the main thread waits for a different thread to run, it can be paused and another thread can take its place until the dependency thread finishes.

The main thread does not take priority over anything else.

Reply:
Was mainly interested in the wording of the last parts `added to the thread pool for jobs`. For me, a thread pool is a place where you go to get new worker threads. So adding the main thread to it is counter intuitive.